### DISCUSSION TITLE
OpenTelemetry .NET and .NET Standard

### DISCUSSION THREAD

**Discussion by alanwest** [open-telemetry/opentelemetry-dotnet#3767] [Announcements]
# OpenTelemetry .NET and .NET Standard

The OpenTelemetry .NET project depends heavily on APIs provided by .NET. Specifically, the APIs provided under `System.Diagnostics.DiagnosticSource`.

With the release of .NET 6, [it was announced](https://github.com/dotnet/announcements/issues/190) that the .NET team will drop out-of-support frameworks for a number of packages including `System.Diagnostics.DiagnosticSource`. This practice will continue with the release of .NET 7. Frameworks no longer supported will include .NET Core 3.1 and .NET 5. Refer to the [.NET download](https://dotnet.microsoft.com/download/dotnet) page to view the end of support dates for each version of .NET.

The core packages offered by OpenTelemetry .NET currently ship a .NET Standard build (i.e., `netstandard2.0` and/or `netstandard2.1`). However, referencing the latest version of OpenTelemetry .NET by an application targeting .NET Core 3.1 or .NET 5 will generate the following build warnings:

> System.Diagnostics.DiagnosticSource doesn't support netcoreapp3.1. Consider updating your TargetFramework to net6.0 or later.

> System.Diagnostics.DiagnosticSource doesn't support net5.0. Consider updating your TargetFramework to net6.0 or later.

These warnings can be suppressed by setting the `SuppressTfmSupportBuildWarnings` MSBuild property.

## .NET Standard builds of OpenTelemetry .NET

OpenTelemetry .NET will continue to ship .NET Standard targets of its packages (i.e., `netstandard2.0` and/or `netstandard2.1`).

Implementation specific target frameworks (e.g., `net8.0`, `net471`) may continue to be added or removed as necessary. Removal of implementation specific target frameworks will always be announced and may occur in minor version releases. For example, in the recent past we removed `net461`, `netcoreapp3.1`, and `net5.0` targets.

### What does support for .NET Standard mean for OpenTelemetry .NET?

We do not declare universal support for .NET Standard because OpenTelemetry .NET is not tested against out-of-support frameworks. Our official support will always be stated in terms of implementation specific target frameworks of .NET (see: https://github.com/open-telemetry/opentelemetry-dotnet#supported-net-versions).

OpenTelemetry .NET can technically be referenced by projects targeting out-of-support frameworks like .NET Core 3.1 and even as far back as .NET Core 2.0. However, there is no guarantee that it will continue to function.

### How does the removal of implementation specific frameworks impact older applications?

Removing implementation specific frameworks may cause older applications to fallback to a .NET Standard build.

For example, if you have an application that targets `net6.0` and you reference the most current version of OpenTelemetry .NET (as of today that is 1.4.0), then you will receive a `net6.0` targeted build of OpenTelemetry. In a future version after .NET 8 is released, for example, we may remove the `net6.0` target and replace it with a `net8.0` target.

You will still be able to upgrade to this later version of OpenTelemetry .NET from your `net6.0` targeted application, however at this point you'll receive a `netstandard2.x` target of OpenTelemetry .NET. Furthermore, you may begin to receive build warnings indicating that your target framework is no longer supported by `System.Diagnostics.DiagnosticSource`.

This can come with some consequences like missing out on performance enhancements or features that were previously available in the old `net6.0` target of OpenTelemetry .NET. We will not intentionally break older applications that fallback to a .NET Standard build, but we cannot make any guarantees.

### Are other implementations of .NET support (e.g., Xamarin or Mono)?

Implementations such as Xamarin and Mono implement `netstandard2.0`. However, it has not been confirmed that OpenTelemetry .NET support these implementations despite currently offering a `netstandard2.0` build (see: https://github.com/open-telemetry/opentelemetry-dotnet/issues/3763). 

If in the future OpenTelemetry .NET declares official support for frameworks like Xamarin or Mono, we may do so either via the `nestandard2.0` build or implementation specific builds (e.g., `xamarin.android`, `net6.0-android`, etc).

**Comment by dpk83** [comment]
> The core packages offered by OpenTelemetry .NET currently ship a .NET Standard build (i.e., netstandard2.0 and/or netstandard2.1). Therefore, OpenTelemetry .NET can technically be consumed by projects targeting out-of-support frameworks like .NET Core 3.1 and even as far back as .NET Core 2.0.

Not true, metrics is broken on 3.1. See  #4204

**Reply by alanwest** [reply]
Apologies for the confusion. I have updated the wording to clarify that OpenTelemetry .NET does not support target frameworks that have reached end of life.

**Reply by sunildatla** [reply]
@alanwest  does this mean opentelemetry for .NET cannot be used for any service or library built on net framework 4.7 ?

**Reply by cijothomas** [reply]
No. Where does it say so?

**Reply by cijothomas** [reply]
see: https://github.com/open-telemetry/opentelemetry-dotnet#supported-net-versions

**Reply by sunildatla** [reply]
Sorry i misread "out-of-support" frameworks to old frameworks and assumed its net framework . Re-reading it made it clear
